File: clone-graph/clone-graph.js
/**
 * // Definition for a Node.
 * function Node(val, neighbors) {
 *    this.val = val === undefined ? 0 : val;
 *    this.neighbors = neighbors === undefined ? [] : neighbors;
 * };
 */

/**
 * @param {Node} node
 * @return {Node}
 */
function cloneGraph(node){
    const map = new Map();
    return clone(node);
    
    function clone(node){
        if(!node) return null;
        if(map.has(node.val)) return map.get(node.val);
        
        var newNode = { val: node.val, neighbors: []};
        map.set(node.val, newNode);
        for (let n of node.neighbors){
            newNode.neighbors.push(clone(n));
        }
        return newNode;
    }
    
}

File: clone-graph/clone-graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    public Node cloneGraph(Node node) {
        HashMap<Node,Node> map=new HashMap<>();
        if(node==null){
            return null;
        }
        map.put(node,new Node(node.val,new ArrayList<>()));
        dfs(map,node);
        return map.get(node);
    }
    public static void dfs(HashMap<Node,Node> map,Node node){
        for(Node neighbor: node.neighbors){
            if(!map.containsKey(neighbor)){
                map.put(neighbor,new Node(neighbor.val,new ArrayList<>()));
                 dfs(map,neighbor);
            }
           
            map.get(node).neighbors.add(map.get(neighbor));
            
        }
    }
}